### ISSUE TITLE
Code generator is no longer usable via `go run`

### ISSUE THREAD

**Issue by twz123** [kubernetes/kubernetes#123933]
### What happened?

The code generator was usable via `go run` up until v1.30.0-alpha.3:

```console
$ go version
go version go1.22.1 linux/amd64
```

```console
$ go run k8s.io/code-generator/cmd/client-gen@v0.30.0-alpha.3 --help
Usage of /tmp/go-build2845286753/b001/exe/client-gen:
[...]
```

Since v1.30.0-beta.0:

```console
$ go run k8s.io/code-generator/cmd/client-gen@v0.30.0-beta.0 --help
go: k8s.io/code-generator/cmd/client-gen@v0.30.0-beta.0 (in k8s.io/code-generator@v0.30.0-beta.0):
        The go.mod file for the module providing named packages contains one or
        more replace directives. It must not contain directives that would cause
        it to be interpreted differently than if it were the main module.
```

### What did you expect to happen?

The code generator works via `go run`.

### How can we reproduce it (as minimally and precisely as possible)?

See above.

### Anything else we need to know?

This may not be a supported way to use the code generator at all, but it worked quite well until now, and I thought I'd bring this to attention. IIUC the offending replace directive came in as part of the workspace support (#123529) and I'm not sure there's anything that can be done about it.

### Kubernetes version

<details>

```console
$ kubectl version
# v1.30.0-beta.0
```

</details>

### Cloud provider

N/A

### OS version

<details>

```console
$ cat /etc/os-release
BUG_REPORT_URL="https://github.com/NixOS/nixpkgs/issues"
BUILD_ID="23.11.20240311.ddcd759"
DOCUMENTATION_URL="https://nixos.org/learn.html"
HOME_URL="https://nixos.org/"
ID=nixos
LOGO="nix-snowflake"
NAME=NixOS
PRETTY_NAME="NixOS 23.11 (Tapir)"
SUPPORT_END="2024-06-30"
SUPPORT_URL="https://nixos.org/community.html"
VERSION="23.11 (Tapir)"
VERSION_CODENAME=tapir
VERSION_ID="23.11"
$ uname -srvmo
Linux 6.6.21 #1-NixOS SMP PREEMPT_DYNAMIC Wed Mar  6 14:48:45 UTC 2024 x86_64 GNU/Linux
```

</details>

### Install tools

N/A

### Container runtime (CRI) and version (if applicable)

N/A

### Related plugins (CNI, CSI, ...) and versions (if applicable)

N/A

**Comment by jiahuif** [comment]
/assign @thockin
Looks like a side effect of the workspace support. Could you take a look?
/triage accepted

**Comment by thockin** [comment]
Ugh.  The workspaces work means that code-generator has a dep (`require`) on apimachinery (https://github.com/kubernetes/code-generator/blob/master/go.mod#L15) and the tooling we use to publish adds that as a `replace` (https://github.com/kubernetes/code-generator/blob/master/go.mod#L40).

It seems redundant to me  - the require and replace are the same version.

@liggitt you know the go deps stuff better than me - couldn't we elide that `replace` when publishing?

**Comment by liggitt** [comment]
makes sense to me, was proposed in https://github.com/kubernetes/publishing-bot/issues/214 https://github.com/kubernetes/publishing-bot/pull/207 but never done

**Comment by liggitt** [comment]
and actually, we ensure there are no replace directives that modify versions any more, so we should be able to drop all replace directives when publishing tags now

**Comment by liggitt** [comment]
opened https://github.com/kubernetes/publishing-bot/pull/415 to resolve this at publish time but I'm not sure how to test publishing-bot changes

**Comment by thockin** [comment]
This should be fix on the next tag!

**Comment by liggitt** [comment]
hmm... didn't work - https://github.com/kubernetes/client-go/blob/v0.31.0-alpha.0/go.mod#L63

**Comment by liggitt** [comment]
@dims dug into logs and we found where the replace directives were getting added back. Followup fix in https://github.com/kubernetes/publishing-bot/pull/416

**Comment by liggitt** [comment]
Now this is fixed, v0.30.0-rc.1 no longer has the replace directives
